components/scheduler/executors.go

```diff
@@ -23,41 +23,472 @@
+
+		if limit != nil {
+			if (*limit)--; *limit <= 0 {
+				break
+			}
+		}
+	}
+	return next.Next, lastAction
+}
+
+func (e taskExecutor) processBackfills(s *Scheduler) {
+	tweakables := e.config.Tweakables()
+
+	limit := tweakables.BackfillsPerIteration
+
+	for len(s.Args.State.OngoingBackfills) > 0 &&
+		limit > 0 &&
+		// use only half the buffer for backfills
+		len(s.Args.State.BufferedStarts) < tweakables.MaxBufferSize/2 {
+		bfr := s.Args.State.OngoingBackfills[0]
+		startTime := timestamp.TimeValue(bfr.GetStartTime())
+		endTime := timestamp.TimeValue(bfr.GetEndTime())
+		next, _ := e.processTimeRange(
+			s,
+			startTime,
+			endTime,
+			bfr.GetOverlapPolicy(),
+			true,
+			&limit,
+		)
+		if next.IsZero() || next.After(endTime) {
+			// done with this one
+			s.Args.State.OngoingBackfills = s.Args.State.OngoingBackfills[1:]
+		} else {
+			// adjust start time for next iteration
+			bfr.StartTime = timestamppb.New(next)
+		}
+	}
+}
+
+func (e taskExecutor) recordAction(s *Scheduler, result *schedpb.ScheduleActionResult, nonOverlapping bool) {
+	s.Args.Info.ActionCount++
+	s.Args.Info.RecentActions = util.SliceTail(append(s.Args.Info.RecentActions, result), RecentActionCount)
+	if nonOverlapping && result.StartWorkflowResult != nil {
+		s.Args.Info.RunningWorkflows = append(s.Args.Info.RunningWorkflows, result.StartWorkflowResult)
+	}
+}
+
+func (e taskExecutor) startWorkflow(
+	s *Scheduler,
+	start *schedspb.BufferedStart,
+	newWorkflow *workflowpb.NewWorkflowExecutionInfo,
+) (*schedpb.ScheduleActionResult, error) {
+	nominalTimeSec := start.NominalTime.AsTime().UTC().Truncate(time.Second)
+	workflowID := newWorkflow.WorkflowId
+
+	// must match AppendedTimestampForValidation
+	workflowID += "-" + nominalTimeSec.Format(time.RFC3339)
+
+	lastCompletionResult, continuedFailure := s.Args.State.LastCompletionResult, s.Args.State.ContinuedFailure
+	if start.OverlapPolicy == enumspb.SCHEDULE_OVERLAP_POLICY_ALLOW_ALL {
+		// ALLOW_ALL runs don't participate in lastCompletionResult/continuedFailure at all
+		lastCompletionResult = nil
+		continuedFailure = nil
+	}
+
+	// TODO(Tianyu): Add callback here
+	req := &workflowservice.StartWorkflowExecutionRequest{
+		WorkflowId:               workflowID,
+		WorkflowType:             newWorkflow.WorkflowType,
+		TaskQueue:                newWorkflow.TaskQueue,
+		Input:                    newWorkflow.Input,
+		WorkflowExecutionTimeout: newWorkflow.WorkflowExecutionTimeout,
+		WorkflowRunTimeout:       newWorkflow.WorkflowRunTimeout,
+		WorkflowTaskTimeout:      newWorkflow.WorkflowTaskTimeout,
+		Identity:                 s.identity(),
+		RequestId:                uuid.NewString(),
+		WorkflowIdReusePolicy:    enumspb.WORKFLOW_ID_REUSE_POLICY_ALLOW_DUPLICATE,
+		RetryPolicy:              newWorkflow.RetryPolicy,
+		Memo:                     newWorkflow.Memo,
+		// TODO(Tianyu): Search is not implemented for now
+		SearchAttributes:     &commonpb.SearchAttributes{IndexedFields: nil},
```

Comment:
Why is this not implemented? Seems like it should be straightforward to add, no?

Reply:
We haven't thought about how search, query etc. work in state machines and I would rather tackle that together in a later PR.